app/src/main/java/de/westnordost/streetcomplete/quests/address/AddAddressStreetForm.kt

```diff
@@ -0,0 +1,179 @@
+
+            if (name.contains(".") || containsAbbreviations) {
+                possibleAbbreviations.add(name)
+            }
+        }
+
+        // TODO, not sure if I should be saving all the names (is there every going to be more than one?)
+        favs.add(javaClass.simpleName,
+                names.first().name, max = 1)
+
+        confirmPossibleAbbreviationsIfAny(possibleAbbreviations) {
+            if(isPlaceName) {
+                applyAnswer(PlaceName(names))
+            } else {
+                applyAnswer(StreetName(names))
+            }
+        }
+    }
+
+    override val otherAnswers = listOf(
+            OtherAnswer(R.string.quest_address_street_no_named_streets) { switchToPlaceName() }
+    )
+
+    @Inject
+    internal lateinit var abbreviationsByLocale: AbbreviationsByLocale
+    @Inject
+    internal lateinit var roadNameSuggestionsDao: RoadNameSuggestionsDao
+
+    @Inject internal lateinit var favs: LastPickedValuesStore<String>
+
+    init {
+        Injector.instance.applicationComponent.inject(this)
+        val lastPickedNames = favs.get(javaClass.simpleName)
+        defaultName = if (lastPickedNames.isEmpty()) {""} else {lastPickedNames.first}
+    }
+
+    override fun setupNameAdapter(data: List<LocalizedName>, addLanguageButton: View): AddLocalizedNameAdapter {
+
+        return AddLocalizedNameAdapter(
+                data, activity!!, getPossibleStreetsignLanguages(),
+                abbreviationsByLocale, getRoadNameSuggestions(), addLanguageButton,
+                defaultName
+        )
+    }
+
+    private fun getRoadNameSuggestions(): List<MutableMap<String, String>> {
+        return roadNameSuggestionsDao.getNames(
+                geometryToMajorPoints(elementGeometry),
+                AddAddressStreet.MAX_DIST_FOR_ROAD_NAME_SUGGESTION
+        )
+    }
+
+    private fun geometryToMajorPoints(geometry: ElementGeometry): List<LatLon> {
+        return when(geometry) {
+            is ElementPolylinesGeometry -> {
+                val polyline = geometry.polylines.first()
+                listOf(polyline.first(), polyline.last())
+            }
+            is ElementPolygonsGeometry -> {
+                // return center and one of nodes from the way constructing area
+                listOf(geometry.center, geometry.polygons.first().last())
+            }
+            is ElementPointGeometry -> {
+                listOf(geometry.center)
+            }
+        }
+
+
+    }
+
+    override fun isRejectingClose() : Boolean {
+        // if the form is complete, we will reject close, unless it is still has only the default name
+        if (isFormComplete()) {
+            if (adapter.localizedNames.size == 1 && adapter.localizedNames.first().name == defaultName) {
+                return false
+            } else {
+                return true
+            }
+        } else {
+            return false
+        }
```

Comment:
You could shorten it to

```suggestion
        return isFormComplete() && !(adapter.localizedNames.size == 1 && adapter.localizedNames.first().name == defaultName
```

Reply:
Also, not sure if the wording of that function is off, but what if the form is not complete but also not empty, should it not also reject closing?

Reply:
The name of the function confuses me a little here too, but I think that in this case the isFormComplete is inherited from AAddLocalizedNameForm which basically says the form is not empty:

```
    // all added name rows are not empty
    override fun isFormComplete() = adapter.localizedNames.isNotEmpty()
            && adapter.localizedNames.all { it.name.trim().isNotEmpty() }
```

So as long as the form is not empty, then it is complete. But I could be wrong, I'm open to improvements.